design/ceph/object/ceph-bucket-notification-crd.md

```diff
@@ -0,0 +1,82 @@
+# Ceph bucket notifications CRD
+
+## Overview
+
+Ceph added support for the bucket notifications feature from Nautilus onwards. It allows to send messages to various endpoints when a new event occurs on a bucket (ref: <https://docs.ceph.com/docs/master/radosgw/notifications/)>
+
+Setup of those notifications are normally done by sending HTTP requests to the RGW, either to create/delete topics pointing to specific endpoints, or create/delete bucket notifications based on those topics.
+
+This functionality eases this process by avoiding to use external tools or scripts. It is replaced by creation of CR definitions that contain all the information necessary to create topics and/or notifications, which the rook operator processes.
+
+## Goals
+
+Creates a CRD for topics and a CRD for notifications, defining all the necessary and optional information for the various endpoints.
+
+Extends the rook operator to handle the CRs that would be submitted by users.
+
+## Implementation
+
+The CR for a topic configuration takes this form:
+
+```yaml
+apiVersion: ceph.rook.io/v1
+kind: CephBucketTopic
+metadata:
+  name: # name of the topic
+  namespace: # namespace where topic belongs
+Spec:
+  endpoint: #(mandatory) URI of an endpoint to send push notification to
+  opaqueData: #(optional) opaque data is set in the topic configuration
+  persistent: #(optional) indication whether notifications to this endpoint are persistent or not (`false` by default)
+  http:
+    verifySSL: #indicate whether the server certificate is validated by the client or not (`true` by default)
+  amqp:
+    ackLevel: # none/routable/broker, optional (default - broker)
+    amqpExchange: # exchanges must exist and be able to route messages based on topics
+  kafka:
+    useSSL: # secure connection will be used for connecting with the broker (`false` by default)
+    caLocation: # this specified CA will be used, instead of the default one, to authenticate the broker
```

Comment:
What is the CA location? The name of a Kubernetes Secret?

Reply:
currently this is passed into the kafka client - should be path to CA file.
to support as a kubernetes secret the we will have to make a change to the RGW code, to use the option where the certificate itself is passed to the library (instead of the location of the file). see: https://github.com/edenhill/librdkafka/blob/master/CONFIGURATION.md
the CRD controller code will have to read the secret and pass it as a parameter in the creation of the topic